File: frontend/src/types/admin/cases.ts
/**
 * Admin Case Management Types
 * Corresponds to backend schemas in app/schemas/admin/cases.py
 */

// === Case Enums ===
export type CaseStatus = 'new' | 'in_progress' | 'waiting' | 'resolved' | 'closed';
export type CaseSeverity = 'low' | 'med' | 'high' | 'critical';
export type SLAStatus = 'safe' | 'warning' | 'critical' | 'breached';

// === Triage Assessment ===
export interface TriageAssessmentSummary {
  id: number;
  risk_score: number;
  severity_level: string;
  confidence_score: number | null;
  risk_factors: string[] | null;
  created_at: string;
}

// === Case Assignment ===
export interface CaseAssignmentSummary {
  id: string; // UUID
  assigned_to: string | null;
  assigned_by: number | null;
  assigned_at: string;
  previous_assignee: string | null;
  reassignment_reason: string | null;
  assignee_role: string | null;
}

// === Case List Item ===
export interface CaseListItem {
  // Base case fields
  id: string;
  status: CaseStatus;
  severity: CaseSeverity;
  user_hash: string;
  session_id: string | null;
  conversation_id: string | null;
  summary_redacted: string | null;
  assigned_to: string | null;
  created_at: string;
  updated_at: string;
  sla_breach_at: string | null;

  // Computed SLA fields
  is_sla_breached: boolean;
  minutes_until_breach: number | null;
  sla_status: SLAStatus;

  // Related counts
  notes_count: number;
  assignments_count: number;

  // Latest triage
  latest_triage: TriageAssessmentSummary | null;
}

// === Case List Response ===
export interface CaseListResponse {
  cases: CaseListItem[];
  total: number;
  page: number;
  page_size: number;
  has_next: boolean;
  has_prev: boolean;
}

// === Case Note ===
export interface CaseNoteItem {
  id: number;
  case_id: string;
  note: string;
  created_at: string;
  author_id: number | null;
}

// === Conversation Message ===
export interface ConversationMessageSummary {
  id: string;
  role: string;
  content: string;
  timestamp: string;
}

// === Case Detail Response ===
export interface CaseDetailResponse {
  // Base case fields
  id: string;
  status: CaseStatus;
  severity: CaseSeverity;
  user_hash: string;
  session_id: string | null;
  conversation_id: string | null;
  summary_redacted: string | null;
  assigned_to: string | null;
  created_at: string;
  updated_at: string;
  sla_breach_at: string | null;

  // Computed SLA fields
  is_sla_breached: boolean;
  minutes_until_breach: number | null;
  sla_status: SLAStatus;

  // Related data
  notes: CaseNoteItem[];
  assignments: CaseAssignmentSummary[];
  triage_assessments: TriageAssessmentSummary[];
  conversation_preview: ConversationMessageSummary[] | null;
}

// === Request Payloads ===
export interface CaseStatusUpdate {
  status: CaseStatus;
  note?: string;
}

export interface CaseAssignmentUpdate {
  assigned_to: string | null;
  reason?: string;
}

export interface CaseNoteCreate {
  note: string;
}

// === Filter & Pagination ===
export interface CaseFilters {
  status?: CaseStatus;
  severity?: CaseSeverity;
  assigned_to?: string;
  unassigned?: boolean;
  sla_breached?: boolean;
  search?: string;
  page?: number;
  page_size?: number;
  sort_by?: 'created_at' | 'updated_at' | 'severity' | 'sla_breach_at';
  sort_order?: 'asc' | 'desc';
}

// === Conversation Response ===
export interface CaseConversationResponse {
  case_id: string;
  conversation_id: string;
  messages: ConversationMessageSummary[];
  total_messages: number;
  conversation_created_at: string | null;
}

File: backend/app/schemas/admin/cases.py
from __future__ import annotations

from datetime import datetime
from typing import Any, List
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class CaseNoteItem(BaseModel):
    id: int
    case_id: str
    created_at: datetime
    author_id: int | None
    note: str


class CaseNotesListResponse(BaseModel):
    items: List[CaseNoteItem]


class CaseNoteCreate(BaseModel):
    note: str


# New schemas for comprehensive case management

class TriageAssessmentSummary(BaseModel):
    """Summary of triage assessment for case list."""
    id: int
    risk_score: float
    severity_level: str
    confidence_score: float | None = None
    risk_factors: list[str] | None = None
    created_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class CaseAssignmentSummary(BaseModel):
    """Summary of case assignment."""
    id: str
    assigned_to: str | None
    assigned_by: int | None
    assigned_at: datetime
    previous_assignee: str | None = None
    reassignment_reason: str | None = None
    assignee_role: str | None = None
    
    model_config = ConfigDict(from_attributes=True)


class CaseListItem(BaseModel):
    """Case item for list view with essential details."""
    id: str
    status: str
    severity: str
    user_hash: str
    session_id: str | None
    conversation_id: int | None
    summary_redacted: str | None
    assigned_to: str | None
    created_at: datetime
    updated_at: datetime
    sla_breach_at: datetime | None
    
    # Computed fields
    is_sla_breached: bool
    minutes_until_breach: int | None
    sla_status: str  # 'safe', 'warning', 'critical', 'breached'
    
    # Related data counts
    notes_count: int = 0
    assignments_count: int = 0
    
    # Latest triage assessment
    latest_triage: TriageAssessmentSummary | None = None


class CaseListResponse(BaseModel):
    """Paginated case list response."""
    cases: list[CaseListItem]
    total: int
    page: int
    page_size: int
    has_next: bool
    has_prev: bool


class ConversationMessageSummary(BaseModel):
    """Summary of conversation message."""
    id: str  # Can be int or string (e.g., "123-user", "123-assistant")
    role: str
    content: str
    timestamp: datetime
    
    model_config = ConfigDict(from_attributes=True)


class CaseDetailResponse(BaseModel):
    """Detailed case view with all related data."""
    id: str
    status: str
    severity: str
    user_hash: str
    session_id: str | None
    conversation_id: int | None
    summary_redacted: str | None
    assigned_to: str | None
    created_at: datetime
    updated_at: datetime
    sla_breach_at: datetime | None
    
    # Computed fields
    is_sla_breached: bool
    minutes_until_breach: int | None
    sla_status: str
    
    # Related data
    notes: list[CaseNoteItem]
    assignments: list[CaseAssignmentSummary]
    triage_assessments: list[TriageAssessmentSummary]
    
    # Conversation preview (if available)
    conversation_preview: list[ConversationMessageSummary] | None = None


class CaseStatusUpdate(BaseModel):
    """Request to update case status."""
    status: str = Field(..., description="New status: new, in_progress, waiting, resolved, closed")
    note: str | None = Field(None, description="Optional note explaining status change")


class CaseAssignmentUpdate(BaseModel):
    """Request to assign/reassign case."""
    assigned_to: str | None = Field(None, description="Agent user identifier to assign case to (None to unassign)")
    reason: str | None = Field(None, description="Optional reason for assignment/reassignment")


class CaseFilters(BaseModel):
    """Filters for case list endpoint."""
    status: str | None = None
    severity: str | None = None
    assigned_to: str | None = None
    unassigned: bool | None = None
    sla_breached: bool | None = None
    search: str | None = None  # Search in user_hash or summary


